include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,454 @@
+ */
+typedef int (*rtc_api_set_time)(const struct device *dev, const struct tm *timeptr);
+
+/**
+ * @typedef rtc_get_time
+ * @brief API for getting RTC time
+ */
+typedef int (*rtc_api_get_time)(const struct device *dev, struct tm *timeptr);
+
+/**
+ * @typedef rtc_alarm_get_supported_fields
+ * @brief API for getting the supported fields of the RTC alarm time
+ */
+typedef int (*rtc_api_alarm_get_supported_fields)(const struct device *dev, uint16_t *mask);
+
+/**
+ * @typedef rtc_alarm_set_time
+ * @brief API for setting RTC alarm time
+ */
+typedef int (*rtc_api_alarm_set_time)(const struct device *dev, uint16_t id, uint16_t mask,
+				      const struct tm *timeptr);
+
+/**
+ * @typedef rtc_alarm_get_time
+ * @brief API for getting RTC alarm time
+ */
+typedef int (*rtc_api_alarm_get_time)(const struct device *dev, uint16_t id, uint16_t *mask,
+				      struct tm *timeptr);
+
+/**
+ * @typedef rtc_alarm_is_pending
+ * @brief API for testing if RTC alarm is pending
+ */
+typedef int (*rtc_api_alarm_is_pending)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_clear_pending
+ * @brief API for clearing RTC alarm pending status
+ */
+typedef int (*rtc_api_alarm_clear_pending)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_set_irq_callback
+ * @brief API for setting alarm IRQ callback
+ */
+typedef int (*rtc_api_alarm_set_irq_callback)(const struct device *dev,
+					      rtc_alarm_irq_callback callback, void *user_data);
+
+/**
+ * @typedef rtc_alarm_enable_irq
+ * @brief API for enabling RTC alarm interrupt
+ */
+typedef int (*rtc_api_alarm_enable_irq)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_alarm_disable_irq
+ * @brief API for disabling RTC alarm interrupt
+ */
+typedef int (*rtc_api_alarm_disable_irq)(const struct device *dev, uint16_t id);
+
+/**
+ * @typedef rtc_set_calibration
+ * @brief API for setting RTC calibration
+ */
+typedef int (*rtc_api_set_calibration)(const struct device *dev, int32_t calibration);
+
+/**
+ * @typedef rtc_get_calibration
+ * @brief API for getting RTC calibration
+ */
+typedef int (*rtc_api_get_calibration)(const struct device *dev, int32_t *calibration);
+
+/**
+ * @brief RTC driver API
+ */
+__subsystem struct rtc_driver_api {
+	rtc_api_set_time set_time;
+	rtc_api_get_time get_time;
+	rtc_api_alarm_get_supported_fields alarm_get_supported_fields;
+	rtc_api_alarm_set_time alarm_set_time;
+	rtc_api_alarm_get_time alarm_get_time;
+	rtc_api_alarm_is_pending alarm_is_pending;
+	rtc_api_alarm_clear_pending alarm_clear_pending;
+	rtc_api_alarm_set_irq_callback alarm_set_irq_callback;
+	rtc_api_alarm_enable_irq alarm_enable_irq;
+	rtc_api_alarm_disable_irq alarm_disable_irq;
+	rtc_api_set_calibration set_calibration;
+	rtc_api_get_calibration get_calibration;
+};
+
+/**
+ * @brief API for setting RTC time
+ *
+ * @param dev Device instance
+ * @param timeptr The time to set
+ *
+ * @return 0 if successful
+ * @return -EINVAL if timeptr is invalid
```

Comment:
What if the time is out of range of the RTC?  I.e. after 2100 or before 2000 or some issue related to how the RTC stores years.  The Linux drivers tend to return `EINVAL` in this case.

Reply:
It shall return -EINVAL, i can expand the description to clarify that invalid time includes time that the device does not support